subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributesContainer.java

```diff
@@ -128,14 +127,24 @@ public <T> AttributeContainer attributeProvider(Attribute<T> key, Provider<? ext
     }
 
     @Override
-    public <T> T getAttribute(Attribute<T> key) {
+    @Nullable
+    public <T> T getAttribute(@Nullable Attribute<T> key) {
         Isolatable<T> isolatable = getIsolatableAttribute(key);
-        return isolatable == null ? null : isolatable.isolate();
+        if (isolatable == null) {
+            return null;
+        }
+
+        T value = isolatable.isolate();
+        if (key != null && value != null && value.getClass() != key.getType() && key.getType() != Object.class) {
+            return isolatable.coerce(key.getType());
+        } else {
+            return value;
+        }
```

Comment:
💭 This is the important change: If we've isolated and didn't get a match for the attribute type, we need to coerce to it.